packages/oauth/oauth-provider/src/router/api-router.ts

```diff
@@ -0,0 +1,899 @@
+      schema == null // No schema means endpoint doesn't accept any input
+        ? async function (req) {
+            req.resume() // Flush body
+            return undefined
+          }
+        : method === 'POST'
+          ? async function (req) {
+              const body = await parseHttpRequest(req, ['json'])
+              return schema.parseAsync(body, { path: ['body'] })
+            }
+          : async function (req) {
+              // @NOTE This should not be necessary with GET requests
+              req.resume().once('error', (_err) => {
+                // Ignore errors when flushing the request body
+                // (e.g. client closed connection)
+              })
+
+              const query = Object.fromEntries(this.url.searchParams)
+              return schema.parseAsync(query, { path: ['query'] })
+            }
+
+    return jsonHandler<C, TReq, TRes>(async function (req, res) {
+      try {
+        // Prevent caching of API routes
+        res.setHeader('Cache-Control', 'no-store')
+        res.setHeader('Pragma', 'no-cache')
+
+        // Prevent CORS requests
+        validateFetchMode(req, ['same-origin'])
+        validateFetchSite(req, ['same-origin'])
+        validateOrigin(req, issuerOrigin)
+        const referer = validateReferer(req, { origin: issuerOrigin })
+
+        // Ensure we are one the right page
+        if (
+          referer.pathname !== '/oauth/authorize' &&
+          referer.pathname !== '/account' &&
+          !referer.pathname.startsWith(`/account/`)
+        ) {
+          throw createHttpError(400, `Invalid referer ${referer}`)
+        }
+
+        // Check if the request originated from the authorize page
+        const requestUri =
+          referer.pathname === '/oauth/authorize'
+            ? await requestUriSchema.parseAsync(
+                referer.searchParams.get('request_uri'),
+              )
+            : undefined
+
+        // Validate CSRF token
+        await validateCsrfToken(req, res)
+
+        // Parse and validate the input data
+        const input = await parseInput.call(this, req)
+
+        // Generate the API response
+        const context = subCtx(this, { input, requestUri })
+        const payload = await buildJson.call(context, req, res)
+
+        return { payload, status: 200 }
+      } catch (err) {
+        onError?.(req, res, err, 'Failed to handle API request')
+
+        // @TODO Rework the API error responses (relying on codes)
+        const payload = buildErrorPayload(err)
+        const status = buildErrorStatus(err)
+
+        return { payload, status }
+      }
+    })
+  }
+}
+
+const EPHEMERAL_COOKIE_OPTIONS: Readonly<CookieSerializeOptions> =
+  Object.freeze({
+    expires: undefined, // session
+    sameSite: 'strict',
+    secure: true,
+    httpOnly: true,
+    path: '/',
+  })
+
+function ephemeralCookieValue(requestUri?: RequestUri) {
+  return requestUri ? `sec-${requestUri}` : 'sec-no-request'
```

Comment:
What does `sec` stand for?

Reply:
Changed to something more explicit (this will likely disappear)